File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function(s, k) {
  const chars = new Map();
  let l = 0;
  let r = 0;
  let result = 1;
  chars.set(s[l], 1);
  while (r < s.length) {
    const maxCount = Math.max(...chars.values());
    if (r - l + 1 - maxCount <= k) {
      result = Math.max(result, r - l + 1);
      r++;
      chars.set(s[r], (chars.get(s[r]) || 0) + 1);
    } else {
      chars.set(s[l], chars.get(s[l]) - 1);
      l++;
    }
  }
  return result;
};

File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
    int[] arr = new int[26];
    int result = 0;
    int max = 0;
    int l = 0;
    for (int r = 0; r < s.length(); r++) {
      arr[s.charAt(r) - 'A']++;
      max = Math.max(max, arr[s.charAt(r) - 'A']);
      if (r - l + 1 - max > k) {
        arr[s.charAt(l) - 'A']--;
        l++;
      }
      result = Math.max(result, r - l+1);
    }
    return result;
    }
}